### PR TITLE
Add advanced redirects example

### PR DESCRIPTION
**Pull request by dominiksipowicz**
### Description

Add redirects example in app-directory.

### Demo URL

Once merged we will provide one.

### Type of Change

- [x] New Example
- [ ] Example updates (Bug fixes, new features, etc.)
- [ ] Other (changes to the codebase, but not to examples)

### New Example Checklist

- [x] 🛫 `npm run new-example` was used to create the example
- [ ] 📚 The template wasn't used but I carefuly read the [Adding a new example](https://github.com/vercel/examples#adding-a-new-example) steps and implemented them in the example
- [ ] 📱 Is it responsive? Are mobile and tablets considered?

### CODE REVIEW

File: app-directory/redirect-with-fallback/app/[...slug]/route.ts [hunk 1]
```diff
@@ -0,0 +1,44 @@
+import { simpleRedirects } from '../../redirects'
+
+export const dynamic = 'error'
+export const dynamicParams = false
```
Comment (lfades) [inline app-directory/redirect-with-fallback/app/[...slug]/route.ts:4]:
```suggestion
```

Comment (lfades) [inline app-directory/redirect-with-fallback/app/[...slug]/route.ts:4]: When `dynamic` is `'error'` the page does not allow dynamic params so this line can be removed.
```suggestion
```

File: app-directory/redirect-with-fallback/redirects.js [hunk 2]
```diff
@@ -0,0 +1,38 @@
+const redirects = [
+  {
+    source: '/simple-redirect-source',
+    destination: '/simple-redirect-destination',
+    permanent: true,
```
Comment (lfades) [inline app-directory/redirect-with-fallback/redirects.js:5]:
```suggestion
    permanent: false,
```

File: app-directory/redirect-with-fallback/app/page.tsx [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+
+export default function Home() {
+  return (
+    <Page className="flex flex-col gap-12">
+      <section className="flex flex-col gap-6">
+        <Text variant="h1">Redirects</Text>
+        <ul>
+          <li>
+            <Link href="/simple-redirect-source">/simple-redirect-source</Link>{' '}
+            - handled in API Route
```
Comment (lfades) [inline app-directory/redirect-with-fallback/app/page.tsx:11]: I think this should be "Route Handler". Also make it an external `<Link>` to https://nextjs.org/docs/app/building-your-application/routing/router-handlers

File: app-directory/redirect-with-fallback/app/[...slug]/route.ts [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+function normalizeSource(source: string): string {
+  return (
+    source
+      .replace(/\\/g, '')
+      .replace(/^\/+/, '')
+      // Trailing slash syntax. Supported by default for the generated routes.
+      .replace(/\{\/\}\?$/, '')
+      .replace(/\/+/g, '/')
+  )
+}
```
Comment (lfades) [inline app-directory/redirect-with-fallback/app/[...slug]/route.ts:44]: Please add comments that explain what this code does.

File: app-directory/redirect-with-fallback/app/[...slug]/route.ts [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+export function GET(_: Request, { params }: { params: Params }): Response {
+  const sourceFromSlug = params.slug.join('/')
+  const redirectInfo = simpleRedirects.find(
+    (r) => normalizeSource(r.source) === sourceFromSlug
+  )
+  if (!redirectInfo) {
+    throw new Error(`Unexpected redirect ${String(params.slug)}`)
+  }
+  const status = redirectInfo.statusCode ?? (redirectInfo.permanent ? 308 : 307)
+  // Response.redirect does not support relative URLs–but browsers do.
```
Comment (lfades) [inline app-directory/redirect-with-fallback/app/[...slug]/route.ts:19]:
```suggestion
  // Response.redirect does not support relative URLs but the `Location` header
  // can be used to indicate a redirect to the browser
```

File: app-directory/redirect-with-fallback/app/complex-redirect-destination/page.tsx [hunk 6]
```diff
@@ -0,0 +1,17 @@
+import { Page, Text, Link, Code, Snippet } from '@vercel/examples-ui'
+import { complexRedirects } from '../../redirects'
+
+export default function Home() {
```
Comment (lfades) [inline app-directory/redirect-with-fallback/app/complex-redirect-destination/page.tsx:4]:
```suggestion
export default function Page() {
```

File: app-directory/redirect-with-fallback/app/page.tsx [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+          </li>
+          <li>
+            <Link href="/complex-redirect-source">
+              /complex-redirect-source
+            </Link>{' '}
+            - handled in <Code>next.config.js</Code>
```
Comment (lfades) [inline app-directory/redirect-with-fallback/app/page.tsx:17]: Be more specific here: Handled by `redirects` in `next.config.js`, and add an external link to the redirects docs.

File: app-directory/redirect-with-fallback/next-env.d.ts [hunk 8]
```diff
@@ -0,0 +1,5 @@
+/// <reference types="next" />
+/// <reference types="next/image-types/global" />
+
+// NOTE: This file should not be edited
+// see https://nextjs.org/docs/basic-features/typescript for more information.
```
Comment (lfades) [inline app-directory/redirect-with-fallback/next-env.d.ts:5]: Add this file to `.gitignore`

File: app-directory/redirect-with-fallback/next.config.js [hunk 9]
```diff
@@ -0,0 +1,11 @@
+const redirects = require('./redirects').complexRedirects
+
+/** @type {import('next').NextConfig} */
+module.exports = {
+  experimental: {
+    appDir: true,
+  },
```
Comment (lfades) [inline app-directory/redirect-with-fallback/next.config.js:7]: This is not needed anymore.
```suggestion
```

File: app-directory/redirect-with-fallback/redirects.d.ts [hunk 10]
```diff
@@ -0,0 +1,9 @@
+export interface Redirect {
+  source: string
+  destination: string
+  permanent?: boolean
+  statusCode?: number
+}
+
+export const simpleRedirects: Redirect[]
+export const complexRedirects: Redirect[]
```
Comment (lfades) [inline app-directory/redirect-with-fallback/redirects.d.ts:9]: Could we move this to JSDoc? consider it optional but I think it makes more sense for this case.

**Review by lfades — CHANGES_REQUESTED** [review]
Looks good. Added multiple comments!

File: app-directory/redirect-with-fallback/next.config.js (hunk 9 above)
Comment (lpalmes) [reply app-directory/redirect-with-fallback/next.config.js:7]: good catch, i've updated next to the latest stable version and removed this

File: app-directory/redirect-with-fallback/redirects.d.ts (hunk 10 above)
Comment (lpalmes) [reply app-directory/redirect-with-fallback/redirects.d.ts:9]: Tried using JSDoc but other ts files were not able to infer the exported types:
```
app/[...slug]/route.ts:1:10 - error TS7034: Variable 'simpleRedirects' implicitly has type 'any[]' in some locations where its type cannot be determined.

1 import { simpleRedirects } from '../../redirects'
```
I've removed `redirects.d.ts` and added JSDoc comments in `redirects.js`
```js
/** @typedef {{
  source: string
  destination: string
  permanent?: boolean
  statusCode?: number
}} Redirect */

/** @type {Redirect[]} */
const simpleRedirects = []
/** @type {Redirect[]} */
const complexRedirects = []
```

**Review by lfades — APPROVED** [review]
LGTM. Remove the `next-env.d.ts` file before merging.

File: app-directory/redirect-with-fallback/turbo.json [hunk 11]
```diff
@@ -0,0 +1,9 @@
+{
+  "$schema": "https://turbo.build/schema.json",
```
Comment (leerob) [inline app-directory/redirect-with-fallback/turbo.json:2]: Is there a reason Turbo is added here?

**Pull request merged**